addons/package_hierarchy/models/package_links.py

```diff
@@ -0,0 +1,185 @@
+                                            "to multiple locations."))
+            elif count > 2:
+                raise ValidationError(_("Links are proposing to move package "
+                                        "to multiple locations."))
+
+        # Make sure all links have child_id
+        if not all(link.child_id for link in self):
+            raise ValidationError(_("Links should always have a child."))
+
+        # Create chains from the links
+        chains = self._return_chains()
+        # Create a dict of links for efficient searching for unlinks later
+        unlinks_from_child = {link.child_id: link for link in self if not link.parent_id}
+
+        # Now need to check these chains against the current reality
+        # NOTE: Currently this is done for a single chain at a time, if we wanted to be
+        # thorough we should check for multiple chains within the same tree at once. This
+        # would only occur for very complex moves so the computational effort may not be
+        # worth it for a 'normal' system.
+        for chain in chains:
+            # Check for self-ancestors
+            current_ancestors = chain[-1]._return_ancestors()
+            if any(True for node in chain if node in current_ancestors):
+                raise ValidationError(_("Proposed link(s) would result in a package loop"))
+
+            # Check the depth of the proposed tree
+            # Check if there is a relevant unlink
+            if unlinks_from_child.get(chain[-1], False):
+                length_above = 0
+            else:
+                length_above = chain[-1]._return_num_ancestors()
+            total_length = length_above + len(chain) + chain[0].depth - 1
+            if total_length > max_package_depth:
+                raise ValidationError(_("Proposed link(s) would cause package depth "
+                                        "to exceed maximum permitted"))
+
+    def _return_chains(self):
+        """Create chains out of links in self.
+        These are returned as a list of lists of records to ensure that they are ordered.
+        """
+        # 1 Get all terminal children and parents
+        # (excluding unlinks as they do not impact the chains)
+        max_package_depth = self.env.user.get_user_warehouse().u_max_package_depth
+
+        links_excluding_unlinks = [link for link in self if link.parent_id]
+        parents = set(link.parent_id for link in links_excluding_unlinks)
+        children = set(link.child_id for link in links_excluding_unlinks)
+        terminal_parents = parents - children
+        terminal_children = children - parents
+
+        # Create a dict of links for efficient searching later
+        link_from_child = {link.child_id: link for link in links_excluding_unlinks}
+
+        # Create a dict to keep track of links that have been checked
+        links_checked = {link: False for link in links_excluding_unlinks}
```

Comment:
What about doing : `links_not_checked = links_excluding_unlinks`
Then in line 165: `links_not_checked -= current_link`
And in line 182: `if links_not_checked:`